ambry-store/src/main/java/com.github.ambry.store/HardDeleter.java

```diff
@@ -111,18 +117,27 @@ public void run() {
     try {
       while (running.get()) {
         try {
-          if (!hardDelete()) {
-            isCaughtUp = true;
-            synchronized (lock) {
+          reentrantLock.lock();
+          while (isPaused()) {
+            try {
+              pauseCondition.await();
+            } catch (InterruptedException e) {
+              logger.warn("InterruptedException thrown while hard delete thread is paused ", e);
+            }
+          }
+          if (running.get()) {
+            if (!hardDelete()) {
+              isCaughtUp = true;
               if (!running.get()) {
                 break;
               }
-              time.wait(lock, hardDeleterSleepTimeWhenCaughtUpMs);
+              time.wait(reentrantLock, hardDeleterSleepTimeWhenCaughtUpMs);
```

Comment:
this might be incorrect. Don't you have to do `synchronized(reentrantLock)` somewhere to be able to do this? 

Related, why not use `pauseCondition()` for this - this is also another kind of pause.